tests/phpunit/tests/functions/isWpVersionCompatible.php

```diff
@@ -12,14 +12,34 @@ class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
 	 * Tests is_wp_version_compatible().
 	 *
 	 * @dataProvider data_is_wp_version_compatible
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
 	 *
 	 * @ticket 54257
 	 *
 	 * @param mixed $required The minimum required WordPress version.
 	 * @param bool  $expected The expected result.
 	 */
-	public function test_is_wp_version_compatible( $required, $expected ) {
-		$this->assertSame( $expected, is_wp_version_compatible( $required ) );
+	public function test_is_wp_version_compatible( $required, $expected, $expect_exception = false ) {
+
+		if ( $expect_exception ) {
```

Comment:
These latest changes provide a false sense of security, as does the test in [this Core changeset](https://core.trac.wordpress.org/changeset/57332).

For example, if you comment out all datasets _except_ those that should throw an exception, and change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will still pass. This is because an exception was correctly detected, but the test also ended as soon as this occurred. This issue also exists with the test in the Core changeset linked above.

Instead, let's consider this approach:

1. Add a new property to the test class:

```php
/**
 * A notice expected during the tests.
 *
 * @var string
 */
private $expected_notice = '';
```
2. Add a `tear_down()` method that resets the property:
```php
/**
 * Resets the `expected_notice` property after each test runs.
 */
public function tear_down() {
	$this->expected_notice = '';
	parent::tear_down();
}
```
3. Adjust the test so that instead of throwing an exception, we note the notice, restore the previous handler, and check that it matches the expected type and string:
```php
/**
 * Tests is_wp_version_compatible().
 *
 * @dataProvider data_is_wp_version_compatible
 * @runInSeparateProcess
 * @preserveGlobalState disabled
 *
 * @ticket 54257
 *
 * @param mixed $required      The minimum required WordPress version.
 * @param bool  $expected      The expected result.
 * @param bool  $expect_notice Whether to expect a notice.
 */
public function test_is_wp_version_compatible_costdev( $required, $expected, $expect_notice = false ) {
	if ( $expect_notice ) {
		set_error_handler(
			function ( $errno, $errstr ) {
				$this->expected_notice = $errno . ' ' . $errstr;
				restore_error_handler();
			},
			E_ALL
		);
	}

	$result = is_wp_version_compatible( $required );

	if ( $expect_notice ) {
		restore_error_handler();

		$this->assertStringStartsWith(
			E_USER_NOTICE . " is_wp_version_compatible(): `$required` Not a valid WordPress version string.",
			$this->expected_notice,
			'The expected notice was not thrown.'
		);
	}

	$this->assertSame( $expected, $result );
}
```

Now, if we change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will correctly fail.

Run a few checks locally to see the issue with the existing test, and how the changes above resolve this.

Reply:
Looks like the discussion has moved away from having a notice. The above can be disregarded in that case. 🙂

Reply:
Un-resolving this as a decision on the notice hasn't finalized.

@aaronjorbin FYI: Take a look at the first comment in this thread regarding an issue with the changeset you referenced and previous commits to this PR that were influenced by it, and an alternative without this issue.